### ISSUE TITLE
C# rule for Any() vs Count/Length

### ISSUE THREAD

**Issue by Djoums** [green-code-initiative/ecoCode#281]
## Use Count/Length instead of Any()/Count()

**Target**: C# / all systems

**Category** : Performance / Energy savings

**Severity** : Warning

On collections that have a determined count/length, such as lists or arrays, use it instead of calling the Linq Any() or Count() methods. This reduces the number of instructions executed by the CPU at runtime.
```csharp
using System.Collections.Generic;
using System.Linq;

public class C
{
    private readonly List<int> Numbers = new List<int> { 1, 2, 3, 4 };

    public bool WithAny() => Numbers.Any(); // Non compliant, Linq call not needed
    
    public bool WithCount() => Numbers.Count != 0; // Compliant, direct access to property
}
```

Using sharplab.io, we can see the generated asm code is much smaller in the 2nd case.

```asm
C.WithAny()
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: push esi
    L0004: push eax
    L0005: mov esi, [ecx+4]
    L0008: lea edx, [ebp-8]
    L000b: mov ecx, esi
    L000d: call dword ptr [0x20fe6b20]
    L0013: test eax, eax
    L0015: je short L0024
    L0017: xor eax, eax
    L0019: cmp dword ptr [ebp-8], 0
    L001d: setne al
    L0020: pop ecx
    L0021: pop esi
    L0022: pop ebp
    L0023: ret
    L0024: mov ecx, esi
    L0026: call dword ptr [0x20fe6b38]
    L002c: jmp short L0020

C.WithCount()
    L0000: mov eax, [ecx+4]
    L0003: cmp dword ptr [eax+8], 0
    L0007: setne al
    L000a: movzx eax, al
    L000d: ret
```

**Comment by dedece35** [comment]
Hi @Djoums,
thank you, interesting idea !
Do you have time to measure the performance in C# ? 
I don't know if a measurement framework exits for C# but something like JMH framework for Java.

I think this issue should be created in "ecocode-csharp" repo, shouldn't ?

**Comment by Djoums** [comment]
Yes, see below (in general, anything available in Java also exist in C# and vice versa 😉)
But in this case the asm code tells everything really.

| Method    | Mean      | Error     | StdDev    | Allocated |
|---------- |----------:|----------:|----------:|----------:|
| WithAny   | 3.4140 ns | 0.0769 ns | 0.0681 ns |         - |
| WithCount | 0.0220 ns | 0.0206 ns | 0.0193 ns |         - |

Although I just found out this rule is now natively available since .Net 8 (November 2023) :
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1860

No need to recreate it but what we can do is make it available to people using an older SDK through our nuget package, and also changing its severity (from suggestion to warning makes sense for ecoCode).

**Comment by dedece35** [comment]
OK. but until now, we needn't to manage rules for different versions of SDK ... if you have ideas, it's quite good :p

**Comment by Djoums** [comment]
Yes, we can reference Microsoft.CodeAnalysis.NetAnalyzers as a transitive package in our package, that way it will apply to the user code base without having to manually install it. This package always contains all the Roslyn analysis rules, including the newest ones, and is backward compatible for all .Net Core SDKs.
We can also use the .globalconfig file in our package to change the desired severity of those rules (but we'll need to review a lot of them 😉)
I've added this point to the technical base project in the C# repository
